File: leetcode/Algorithms/139.WordBreak/Solution.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        table = [False] * (len(s) + 1)
        table[0] = True
        word_set = set(wordDict)
        for i in range(1, len(s) + 1):
            for j in range(0, i):
                if table[j] and word_set.__contains__(s[j: i]):
                    table[i] = True
        return table[len(s)]

File: leetcode/Algorithms/139.WordBreak/Solution.java
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        boolean[] visited = new boolean[s.length() + 1];

        int length = s.length();
        while (!queue.isEmpty()) {
            int index = queue.poll();
            if (index == length) return true;
            for (String str : wordDict) {
                if (str.length() > length - index) continue;
                if (str.equals(s.substring(index, index + str.length())) && !visited[index + str.length()]) {
                    queue.add(index + str.length());
                    visited[index + str.length()] = true;
                }
            }
        }
        return false;
    }
}
